### DISCUSSION TITLE
How to reset the DB?

### DISCUSSION THREAD

**Discussion by Aprillion** [mswjs/data#51] [Ideas]
When used for testing, we will want to make the tests independent, so some mechanism should exist for resetting all data to initial state (probably outside of @mswjs/data implementation, just some documented recipe).

I would imagine most people who need independent tests already use:

```js
afterEach(() => server.resetHandlers())
```

How to hook into that?

**Comment by kettanaito** [comment]
Thanks for raising this, it's a brilliant point.

## Responsibility

I think this functionality should come from the MSW+data integration. The data package itself shouldn't care about resetting its state, as it's not something you usually do with a real database. 

The request handlers generating API (#11) may take care of that. There's currently no means to hook into MSW `.resetHandlers() ` method, but that can be added as a new life-cycle event:

```js
serverOrWorker.on('internal:reset-handlers', () => {...})
```

```js
import { setupWorker } from 'msw'
import { factory, toHandlers } from '@mswjs/data'

const db = factory({...})

const { handlers, attachEvents } = toHandlers(db)
const worker = setupWorker(...handlers)

// Listens to "internal:reset-handlers" event and resets the database.
attachEvents(worker)
```

_That_ is an incredibly confusing API, and something we can brainstorm on. 

## Reset value

Question: what should DB be reset to? When you model your DB you provide the map of models. It's only after that you can create entities (data) based on those models. 

In other words, what do you expect here to happen:

```js
// Creating a DB, clean.
const db = factory({...})

afterEach(() => server.resetHandlers())

it('test', () => {
  // Request that creates an entity in the DB.
  fetch('/user', { method: 'POST' })
})

it('another test', () => {
  // What entities do you expect the DB to have here?
})
```

I expect your answer to be "no entities". In that case you can use an existing `.clear()` method on the DB:

```js
afterEach(() => {
  server.resetHandlers()
  db.clear()
})
```

> I acknowledge this still should come built-in with #11 so calling `.resetHandlers` would be enough. 

This also comes down to the development pattern you choose:

- One DB per all tests.
- One DB per test.

As it's generally not a good idea for the data to persist between tests, you may want to choose the section option and call `db.clear()` if necessary.

## Semantics

I'm concern with the semantics of connecting `.resetHandlers` method and any DB-related actions. 

What `.resetHandlers` represents is: let's throw away any runtime handlers you might have defined. It's a strong assumption to say that those runtime handlers might have affected your DB. For example, you can append handlers that don't communicate with the DB at all.

**Comment by marcosvega91** [comment]
I add a ittle hint here that I think is important for testing purposes. After the clear is important to have the migration feature.

**Comment by Aprillion** [comment]
> I expect your answer to be "no entities".

Actually, I would be tempted to populate the initial data like following:

```js
// src/mocks/db.js
import initialUsers from "./users.csv"

const db = factory({...})
db.user.createMany(initialUsers)

export {db}
```

or even to have some `const db = factoryFromPostgresDump(filename)` in the future for easy reproduction of bugs by taking a snapshot of production database 🤷‍♂️

So the DB reset before each test would be equivalent to restoring the DB from a dump, not just `.clear()` but also some `.reinitialize()` 🤔 

Hmm, probably the `factory` would need some concept of initial data for this to work...

Or maybe some `const snapshotRef = db.snapshot()` and `db.restore(snapshotRef)`?

**Reply by kettanaito** [reply]
> Hmm, probably the factory would need some concept of initial data for this to work...

Exactly. There must be a point to revert to. 

I like the idea of snapshots, but they may raise more questions than they solve, for instance:

1. Can I have multiple snapshots?
2. Where are the snapshots stored?

As with all our APIs, we prioritize simplicity and scoped responsibility. I don't think the `db` instance should be responsible for anything else than populating the DB based on the models you've described. It is the lack of options that makes an intuitive API.

**Reply by Aprillion** [reply]
1. yes
2. as a return value that can be assigned to a variable? (and/or saved to disk outside of the library)

**Comment by kettanaito** [comment]
## Suggestion: Snapshots

As @Aprillion has suggested, we may use a concept of snapshots—images of the database through time to reset to.

```js
import { factory, snapshot } from '@mswjs/data'

const db = factory({...})

db.user.create()
db.post.create()
db.tags.createMany()

// Take a named snapshot of all the entities in the DB.
snapshot.take(db, 'my-snapshot')

// Tests happen here...

// Restores the DB to the previously taken snapshot. 
snapshot.reset(db, 'my-snapshot')
```

### Behavior

- Snapshot acts as a singleton that accumulates DB images (i.e. `Map<SnapshotName, SerializedDb>`).
- Calling `snapshot.take` an an empty DB throws an exception.
- Restoring a snapshot with an unknown name throws an exception.

**Reply by Aprillion** [reply]
sounds like a good interface for most cases... but if the library catches on, I can see feature requests to support static typing of the snapshot names 

which can be worked around by using variables, but compare to an alternative API suggestion:

```ts
const resetToSnapshot01 = snapshot(db)

// Tests...

resetToSnapshot01()
```

**Reply by kettanaito** [reply]
That API looks okay to me. You'd have to carry the restoration method around, but that's a small price to pay for preventing snapshot naming weirdness.

**Comment by kettanaito** [comment]
## Update

In v0, you could use `.drop()` to remove all the entries in your factory.

Since v1.0, you can use `.clear()` on the collection to remove any records it has:

```ts
const users = new Collection({ schema })

// Delete any records this collection has.
users.clear()

users.all() // []
```

We currently don't hany any snapshot capabilities. Please let me know if those are still needed, I'd be happy to discuss them in more detail.
